In [ ]:
pip install sentence-transformers

Step 1: Loading the embedding model

In [ ]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('all-MiniLM-L6-v2')
print('Model loaded')

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:124: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Model loaded


Step 2: Creatnig Documents

In [ ]:
documents=[
    "Built a production RAG pipeline using LangChain, FAISS, and Claude via AWS Bedrock to enable semantic search across 100K+ compliance documents.",
    "Extended RAG platform with multi-step query rewriting, hybrid retrieval, and agentic workflows using LangGraph.",
    "Evaluated answer quality using LLM-as-judge evaluation frameworks on AWS Bedrock.",
    "Built streaming fraud scoring pipelines using Apache Kafka and Spark Structured Streaming serving predictions via SageMaker endpoints.",
    "Optimized GenAI inference through semantic caching, model benchmarking, and prompt engineering on AWS Bedrock."
]

print(f"We have {len(documents)} documents")
print("\nFirst document:",documents[0])
print("Resume bullets encoded")

We have 5 documents

First document: Built a production RAG pipeline using LangChain, FAISS, and Claude via AWS Bedrock to enable semantic search across 100K+ compliance documents.
Resume bullets encoded


Step 3: Converting documents to vectors

In [ ]:
import numpy as np

# Converting documents to vectors
document_vectors = model.encode(documents)

print(f"Shape: {document_vectors.shape}")
print("\nFirst document vector (first 10 numbers):", document_vectors[0][:10])

Shape: (5, 384)

First document vector (first 10 numbers): [-0.04143848  0.05917067  0.00909833 -0.029286    0.0266975  -0.06166922
 -0.05612318  0.00669392 -0.04034173 -0.00028462]


shape: (5, 384)
5 = no of documents
384 = number of dimensions in each vector, (each sentence is represented as 384 numbers)

Step 4: Convert a question to a vector and find the most similar document

In [ ]:
# question asked by an analyst
question = "Experience with LLMs and retrieval augmented generation systems"

# converting question to vector
question_vector = model.encode([question])

# Calculating similarity between question and every document
similarities = np.dot(document_vectors, question_vector.T).flatten()

print("Similarity scores:")
for i, score in enumerate(similarities):
  print(f" [{score:.3f}] {documents[i]}")

best_idx = np.argmax(similarities)
print(f"\nMost relevant document: ", documents[best_idx])

Similarity scores:
 [0.228] Built a production RAG pipeline using LangChain, FAISS, and Claude via AWS Bedrock to enable semantic search across 100K+ compliance documents.
 [0.341] Extended RAG platform with multi-step query rewriting, hybrid retrieval, and agentic workflows using LangGraph.
 [0.350] Evaluated answer quality using LLM-as-judge evaluation frameworks on AWS Bedrock.
 [0.033] Built streaming fraud scoring pipelines using Apache Kafka and Spark Structured Streaming serving predictions via SageMaker endpoints.
 [0.332] Optimized GenAI inference through semantic caching, model benchmarking, and prompt engineering on AWS Bedrock.

Most relevant document:  Evaluated answer quality using LLM-as-judge evaluation frameworks on AWS Bedrock.


In [ ]:
question_two = "Real time data streaming and ML model deployment at scale"

question_two_vector = model.encode([question_two])
similarities = np.dot(document_vectors, question_two_vector.T).flatten()

print("Relevance scores:")
for i, score in enumerate(similarities):
  print(f" [{score:.3f}] {documents[i]}")

best_idx = np.argmax(similarities)
print(f"Most relevant document: ", documents[best_idx])

Relevance scores:
 [0.190] Built a production RAG pipeline using LangChain, FAISS, and Claude via AWS Bedrock to enable semantic search across 100K+ compliance documents.
 [0.116] Extended RAG platform with multi-step query rewriting, hybrid retrieval, and agentic workflows using LangGraph.
 [0.064] Evaluated answer quality using LLM-as-judge evaluation frameworks on AWS Bedrock.
 [0.386] Built streaming fraud scoring pipelines using Apache Kafka and Spark Structured Streaming serving predictions via SageMaker endpoints.
 [0.261] Optimized GenAI inference through semantic caching, model benchmarking, and prompt engineering on AWS Bedrock.
Most relevant document:  Built streaming fraud scoring pipelines using Apache Kafka and Spark Structured Streaming serving predictions via SageMaker endpoints.


# RAG

This notebook demonstrates the core of Retrieval Augumented Generation(RAG):

**R - Retrieval:** We convert the question to a vector and search for the most simmilar document using dot product similarity. We retrieve the most relevant context.

**A - Augemented:** Instead of just sending the question to an LLM, we AUGMENT the prompt with the retrieved document as context. This grounds the LLM's answer in real data.

**G - Generation:** The LLM generates an answer based on the retrieved context, not from its training data. This prevents hallucination.

Without retrieval , the LLM gusses. With retrieval, it answers from facts.

---

## Why I used a all-MiniLM-L6-v2?

MiniLM =. Mini Language Model -  a smaller, faster version of larger models.

L6 = 6 transformer layers(larger models have 12 or more).

v2 = version 2